Support Sparkfun ICM20948 Breakout board (SPI only, I2C TBD)

atinm opened the pull request:
I have only tested SPI and 9DOF, I didn't try I2C at all. I also couldn't make the Adafruit ICM20948 board do SPI, there seems to be a hardware limitation. I pulled in the whole Sparkfun code with some minor edits (include strings.h, uncomment DMP usage) - it is MIT licensed. I don't really like wholesale copying of external repos and locally I use submodules, but I don't know how that would work with the Arduino IDE and pulling in madflight as a library.

I also had to change computeAngles() for NED and flip the sign so that clockwise around Z would be positive. I am not sure what that would do to other IMUs, it seems to be required to have clockwise be positive.

atinm committed:
Added ICM20984

atinm committed:
Added Sparkfun ICM20948 breakout board (SPI only, I2C TBD)

I couldn't get the Adafruit ICM20948 board to support SPI and I
didn't want to use I2C. This code is only tested with SPI and
9DOF.

qqqlab commented:
Hi @atinm, many thanks for your PRs!

I'm currently in holiday mode, and it will take some time before I'll return to my computer keyboard. My comments below are based only on reading your code.

I'm not a big fan of copying libs either, but as Arduino IDE has no option to specify a specific lib version I think copying is the way to go to make sure madflight will compile in the future.

Could you please trim down the copied sparkfun files: delete the folders: examples, .vscode, img and delete the files in the lib root except licence.md and readme.md

For the Arduino IDE we probably need to move the files in SparkFun_ICM-20948_ArduinoLibrary/src one level down to SparkFun_ICM-20948_ArduinoLibrary, or make the util folder a subfolder of src. Otherwise #include "util/ICM_20948_C.h" in src/imu/ICM20948/SparkFun_ICM-20948_ArduinoLibrary/src/ICM_20948.h will fail (I think...)

atinm commented:
I am going to try building it exactly as I have it rather than via submodules and see if it works and fix any compile issues.

atinm committed:
Merge branch 'main-upstream' into atinm/icm20948-imu

atinm committed:
Copied files to top level from Sparkfun's subdirectory and updated README.md

Just use the files necessary for Madflight. Updated Readme to point to the
main repository from Sparkfun and for the documentation and license.

qqqlab commented:
Removing the folders is required for Arduino, as Arduino compiles all source files in the src folder. At least that is what I thought, I'm surprised to see that for R2040 it compiles...

But anyway, these examples (and the other folders) are not required for madflight. If somebody wants to dive into this particular driver, they can look at the sparkfun repo.

With ESP32 we get this error: imu.h:35:5: error: 'PinStatus' does not name a type.

Apparently PinStatus is not defined for this Arduino flavor. Probably use: `bool has_interrupt_rising _edge = false;` to make it portable.

Also, I would make has_interrupt_rising_edge a property of ImuGizmo, and not of ImuConfig. I.e. just like has_sensor_fusion, which is also not user configurable, both properties are set by the driver.

atinm committed:
Added back the config for imu_int_mode for ICM20948

atinm committed:
Use uint8_t for interrupt_mode

atinm committed:
Handle when we aren't using MBED or RP2040

atinm commented:
@qqqlab actually, not a bad idea removing it from config, but I think I still have to deal with the call to attachInterrupt taking either a PinStatus or an int depending on which board we are using.

atinm committed:
Moved interrupt_mode from config to gizmo

atinm commented:
Done - moved the interrupt mode to gizmo, default to RISING.

qqqlab commented:
Wow, starts to look really good! Thanks!

Apparently I was not clear enough, this is what I was thinking of:

if(has_interrupt_rising _edge) {
  attachInterrupt(digitalPinToInterrupt(interrupt_pin), _imu_ll_interrupt_handler, RISING); 
}else{
  attachInterrupt(digitalPinToInterrupt(interrupt_pin), _imu_ll_interrupt_handler, FALLING);
}

Will work for any arduino flavor, and we get rid of #if defined(ARDUINO_ARCH_MBED) || defined(ARDUINO_ARCH_RP2040).

atinm committed:
Use bool to pass interrupt mode to _imu_ll_interrupt_setup

atinm commented:
Done - pass along a bool (still need to change _imu_ll_interrupt_handler's prototype as it doesn't have gizmo available), and removed the #if defined ARDUINO_ARCH_MBED etc.

atinm committed:
Added some comments about the quaternion math and fixed yaw radian calculation

atinm committed:
Make the IMU interrupt check a loop rather than a die for IMUs that take a while to come up

qqqlab commented on src/ahr/ahr.cpp:
This will break existing code, should not change this.

Probably has to do with different mag and gyro/acc axis orientation, see page 83 of icm20946 datasheet. See also mpu9250 driver, as this chip has the same issue.

atinm replied:
I thought I had added the minus sign and then realized it was wrong! Will remove.

qqqlab commented on src/imu/ICM20948/ImuGizmoICM20948.cpp:
Can we move this file to src/imu please? (so that the ICM20948 folder contains only external driver files)

qqqlab commented on src/madflight.h:
Please remove 

    // effectively the same thing, but lets us recover if the IMU is slow to start generating interrupts.

makes sense in diff, but not in final code

qqqlab commented on src/cfg/cfg.h:
don't you want to add ahr_gizmo mf_IMU?

(also in ahr.cpp)

atinm replied:
Weird, I know I had this but it didn't make the PR. Will update.

atinm committed:
Added mf_IMU for AHR and fixed sign

atinm committed:
Removed extra comment

atinm committed:
Moved ImuGizmoICM20948.cpp to src/imu

Since the code is split into .h and .cpp, imu.cpp does not
run into register naming conflicts anymore.

atinm committed:
Add src/imu/ImuGizmoICM20948.cpp since move

atinm commented:
I might try one more thing by adding a config option for magnetometer so I try it with both on and off (don’t need magnetometer for FPV) as the ICM20948 supports both 9DOF and 6DOF sensor fusion and it should be configurable rather than hardcoded which to use.

atinm committed:
Allow imu_has_mag config option

Some IMUs like the ICM20948 allow either 9DOF or 6DOF - allow
setting via config whether or not to use magnetometer.

atinm commented on src/ahr/ahr.cpp:
```diff
@@ -43,9 +44,20 @@ int Ahr::setup() {
   cfg.printModule(MF_MOD);
 
   B_gyr = lowpass_to_beta(config.gyrLpFreq, config.pimu->getSampleRate());
+  if (isnan(B_gyr)) {
```
We have a race - we are setting up the ahr before we have set up the IMU in madflight.h so getSampleRate() might be 0. I am not sure what to do about this. I think ahr.setup() should be called after IMU is ready.

qqqlab replied:
I was actually looking at this yesterday, but did not think of the sample rate issue so thought it would be fine.

So yes: first do imu.setup, then ahr.setup, then connect imu interrupt.

And then we can also add a error message in ahr.setup when IMU is selected, but the imu does not support DMP (has_sensor_fusion==false)

atinm replied:
Done

atinm committed:
Fix config->has_mag - it is a bool

atinm committed:
Move AHR setup after IMU setup, and separate interrupt setup

atinm committed:
Die if using mf_IMU on AHR with an IMU that does not have sensor fusion

qqqlab commented on src/ahr/ahr.cpp:
Let's put this in a new PR - I think this will need some additional considerations... 

Some (random) thoughts:

Rather add a new imu_gizmo mf_ICM20948_6DOF than a new parameter. (I try to keep number of parameters as low as possible)

If adding a parameter:

New config params should be added at the end of the array - this allows users to upgrade without eeprom corruption.

I would use value 0 or 1 (not <=0 or 1)

Default value 1 makes more sense to me - use mag if we have it.

Rather name the param and related vars imu_use_mag, and reserve the "has" properties for driver capabilities. 

Not consistent with current 9dof drivers...

atinm replied:
Yeah, not consistent with existing drivers - made it use_mag and moved to bottom. Only really used by ICM20948 for now, could be used by MPU9150 if it supports 6DOF. But left other others alone to have minimal code changes.

atinm replied:
In the meantime, my quadcopter crashed and burned badly - need to debug it!

atinm replied:
I seemed to have lost some code when I split Icm20948 and ibus code so putting it back - related to calibration and quaternions

atinm committed:
Moved imu_use_mag to bottom and pass it along to the IMUs

Currently only ICM20948 supports actuall using this to enable or
disable using its magnetometer, others are just setting has_mag
to true or false but, for example, MPU9150 could possibly use
this as well if it supports 6DOF.

For now left other IMUs as is so as to have minimal code changes.

atinm committed:
Set sample rate to 225 Hz which is the best ICM20948 can do

atinm committed:
Do bias correction on the quaternions as well

The quaternions need bias correct, and copying into AHRS.
Also added cli commands to show quaternions for IMU and AHRS.

atinm committed:
Fix more Quaternion calculations and calibration

Added calibration of the quaternions for sensor fusion IMUs.
Also added pahq and pquat for showing the quaternions from the
AHR and the IMU. And finally, added a flipYaw() virtual function
to flip the sign of yaw after computeAngles() for IMUs like the
ICM20948 that are in LH Android ENU and just quaternion manipulation
isn't getting the yaw to the correct sign in computAngles().

atinm committed:
Fix quaternion align and remove hack for flipYaw()

qqqlab commented:
Hi @atinm, please let me know when you've added all the features you want to have for a PR, and you're confident to merge the PR, then I'll do a final code review. It's more efficient for me than doing many partial reviews, thanks. 

Specific PR related discussions and questions are always welcome of course.

atinm commented:
Hold off while I get my copter flying with this new code!
